src/kmeans.jl

```diff
@@ -1,91 +1,115 @@
+                 weights::Union{Nothing, AbstractVector{TW}}=nothing, # in: sample weights (n)
+                 maxiter::Integer=_kmeans_default_maxiter, # in: maximum number of iterations
+                 tol::Real=_kmeans_default_tol,            # in: tolerance of change at convergence
+                 display::Symbol=_kmeans_default_display,  # in: level of display
+                 distance::SemiMetric=SqEuclidean()        # in: function to compute distances
+                 ) where {TX<:Real,TC<:Real,TW<:Real}
+    p, n = size(X)
+    p2, k = size(centers)
+
+    p == p2 || throw(DimensionMismatch("Inconsistent array dimensions " *
+                                       "for `X` and `centers`."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
+    if !(weights === nothing)
+        length(weights) == n || throw(DimensionMismatch("Incorrect length " *
+                                                        "of weights."))
+    end
 
     assignments = Vector{Int}(undef, n)
-    costs = Vector{T}(undef, n)
     counts = Vector{Int}(undef, k)
-    cweights = Vector{Float64}(undef, k)
 
-    _kmeans!(X, conv_weights(T, n, weights), centers,
-             assignments, costs, counts, cweights, round(Int, maxiter), tol,
+    TC2 = promote_type(TX, TC)
+    weights === nothing || (TC2 = promote_type(TC2, TW))
+    # corner case where that's still not an AbstractFloat
+    TC2 = float(TC2)
+
+    _kmeans!(X, weights, convert(Matrix{TC2}, centers),
+             assignments, counts, round(Int, maxiter), tol,
              display_level(display), distance)
 end
 
-function kmeans(X::AbstractMatrix{<:Real}, k::Int;
-                weights=nothing, init=_kmeans_default_init,
-                maxiter::Integer=_kmeans_default_maxiter,
-                tol::Real=_kmeans_default_tol,
-                display::Symbol=_kmeans_default_display,
-                distance::SemiMetric=SqEuclidean())
 
-    m, n = size(X)
+function kmeans(X::AbstractMatrix{TX}, # in: sample matrix (p x n) columns = observations
+                k::Int;                # in: number of centers
+                weights::Union{Nothing, AbstractVector{TW}}=nothing, # in: sample weights (n)
+                init=_kmeans_default_init,                # in: initialization algorithm
+                maxiter::Integer=_kmeans_default_maxiter, # in: maximum number of iterations
+                tol::Real=_kmeans_default_tol,            # in: tolerance  of change at convergence
+                display::Symbol=_kmeans_default_display,  # in: level of display
+                distance::SemiMetric=SqEuclidean()        # in: function to calculate distance with
+                ) where {TX<:Real,TW<:Real}
+    p, n = size(X)
     (2 <= k < n) || error("k must have 2 <= k < n.")
+
+    TC = float(TX)
+    centers = Matrix{TC}(undef, p, k)
+
     iseeds = initseeds(init, X, k)
-    centers = copyseeds(X, iseeds)
+    copyseeds!(centers, X, iseeds)
```

Comment:
Just a small style suggestion:
```suggestion
    centers = copyseeds!(Matrix{TC}(undef, p, k), X, iseeds)
```
would avoid uninitialized `centers` hanging around.

Reply:
ok, I'll look at this with your other comment of maybe removing `copyseeds` 🙂